### PR TITLE
#2301 switch shunt

### PR DESCRIPTION
**Pull request by jlbouchot**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [ ] unit tests and non regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.extvar [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, you can obtain one at http://mozilla.org/MPL/2.0/.
+    SPDX-License-Identifier: MPL-2.0
+
+    This file is part of Dynawo, an hybrid C++/Modelica open source suite of
+    simulation tools for power systems.
+-->
+<external_variables xmlns="http://www.rte-france.com/dynawo">
+  <variable id="u1" type="discrete"/>
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.extvar:15]: Are u1 and u2 necessary discrete in your case ?

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, you can obtain one at http://mozilla.org/MPL/2.0/.
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, a hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+block Switch "Switches the inputs based on an xor of two booleans, one of which may change during the simulation"
+  import Modelica;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:16]: Useless here

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  import Dynawo.Connectors;
+  import Modelica.Blocks.Interfaces;
+  import Modelica.Blocks.Icons.Block;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:19]: What is it for ?

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  Connectors.ZPin doSwitch(value(start = -1)) "Whether a switch should occur or not. If off, then input 1 is returned, else input 2" annotation(Placement(visible = true, transformation(origin = {-110, 70}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {-160, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+
+  discrete Interfaces.RealInput u1;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:23]: Maybe a simple comment for u1 and u2.

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  discrete Interfaces.RealInput u2;
+
+  Interfaces.RealOutput y "Output signal connector" annotation (Placement(
+        transformation(extent={{100,-10},{120,10}})));
+
+  parameter Boolean isCapa "In case the expected behaviour should be inverted from the default one. This is typically the case if the output is processed by a capacitor even though the implementation is done with a self in mind.";
+
+equation
+
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:32]: No line between equation and the first equation

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+* All rights reserved.
+* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, you can obtain one at http://mozilla.org/MPL/2.0/.
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, a hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+block Switch "Switches the inputs based on an xor of two booleans, one of which may change during the simulation"
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:15]: Maybe Switch is too generic, but I don't have anything better in mind :(

**Review by cossartque — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.extvar (hunk 1 above)
Comment (jlbouchot) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.extvar:15]: Yes! 
The documentation / comments in the .mo have been modified to make it more obvious.

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo (hunk 4 above)
Comment (jlbouchot) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:23]: Done

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, a hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+block Switch "Switches two discrete inputs based on an xor of two booleans, one of which may change during the simulation"
+  import Dynawo.Connectors;
+  import Modelica.Blocks.Interfaces;
+
+  Connectors.ZPin doSwitch(value(start = -1)) "Whether a switch should occur or not. The filtering is done based on the discrete values of an ACMC: 0 undefined, 1 open, 2,3,4,5 closed" annotation(Placement(visible = true, transformation(origin = {-110, 70}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {-160, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
```
Comment (marcochiaramello) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:19]: please remove any reference to ACMC

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo (hunk 7 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:19]: Also, why is it not a discrete Real Input like u1 and u2 ?

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+  discrete Interfaces.RealInput u1 "One of the discrete inputs";
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:21]: Please add also an annotation for u1 and u2 + review the placement of the blocks in the diagram view and icon view.

File: dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  discrete Interfaces.RealInput u2 "The other discrete input";
+
+  Interfaces.RealOutput y "Output signal connector" annotation (Placement(
+        transformation(extent={{100,-10},{120,10}})));
+
+  parameter Boolean invertBehaviour "In case the expected behaviour should be inverted from the default one.";
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/NonLinear/Switch.mo:27]: Should be put before the inputs and connectors

**Review by cossartque — CHANGES_REQUESTED** [review]

**Comment by cossartque** [comment]
Also, please rebase your branch

**Review by cossartque — CHANGES_REQUESTED** [review]

**Review by marcochiaramello — APPROVED** [review]

**Review by cossartque — APPROVED** [review]

**Pull request merged**